libproto/src/lib.rs

```diff
@@ -68,83 +60,79 @@ impl TxResponse {
-    pub const VERIFY_TX_RESP: u16 = 9;
-    pub const VERIFY_BLK_REQ: u16 = 10;
-    pub const VERIFY_BLK_RESP: u16 = 11;
-    pub const BLOCK_TXHASHES: u16 = 12;
-    pub const BLOCK_TXHASHES_REQ: u16 = 13;
-    pub const NEW_PROOF_BLOCK: u16 = 14;
-    pub const BLOCK_TXS: u16 = 15;
-    pub const RICH_STATUS: u16 = 16;
-    pub const EXECUTED_RESULT: u16 = 17;
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum SubModules {
+    Jsonrpc,
+    Net,
+    Chain,
+    Consensus,
+    Auth,
+    Executor,
+    Unknown,
 }
 
-#[derive(Debug)]
-pub enum MsgClass {
-    EMPTY,
-    REQUEST(Request),
-    RESPONSE(Response),
-    HEADER(BlockHeader),
-    BLOCK(Block),
-    STATUS(Status),
-    VERIFYTXREQ(VerifyTxReq),
-    VERIFYTXRESP(VerifyTxResp),
-    VERIFYBLKREQ(VerifyBlockReq),
-    VERIFYBLKRESP(VerifyBlockResp),
-    BLOCKTXHASHES(BlockTxHashes),
-    BLOCKTXHASHESREQ(BlockTxHashesReq),
-    BLOCKWITHPROOF(BlockWithProof),
-    BLOCKTXS(BlockTxs),
-    MSG(Vec<u8>),
-    RICHSTATUS(RichStatus),
-    SYNCREQUEST(SyncRequest),
-    SYNCRESPONSE(SyncResponse),
-    EXECUTED(ExecutedResult),
+pub const SUBMODULES_UNKNOWN: SubModules = SubModules::Unknown;
+
+impl fmt::Display for SubModules {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                // Please use the same rules to name the string
+                &SubModules::Jsonrpc => "jsonrpc",
+                &SubModules::Net => "net",
+                &SubModules::Chain => "chain",
+                &SubModules::Consensus => "consensus",
+                &SubModules::Auth => "auth",
+                &SubModules::Executor => "executor",
+                &SubModules::Unknown => "unknown",
+            }
+        )
+    }
 }
 
-pub fn key_to_id(key: &str) -> u32 {
-    if key.starts_with("jsonrpc") {
-        submodules::JSON_RPC
-    } else if key.starts_with("net") {
-        submodules::NET
-    } else if key.starts_with("chain") {
-        submodules::CHAIN
-    } else if key.starts_with("consensus_cmd") {
-        submodules::CONSENSUS_CMD
-    } else if key.starts_with("consensus") {
-        submodules::CONSENSUS
-    } else if key.starts_with("auth") {
-        submodules::AUTH
-    } else if key.starts_with("executor") {
-        submodules::EXECUTOR
-    } else {
-        0
+impl<'a> From<&'a str> for SubModules {
+    fn from(s: &'a str) -> SubModules {
+        let parts: Vec<&str> = s.split('.').collect();
+        if parts.len() > 1 {
+            match parts[0] {
+                "jsonrpc" => SubModules::Jsonrpc,
+                "net" => SubModules::Net,
+                "chain" => SubModules::Chain,
+                "consensus" => SubModules::Consensus,
+                "auth" => SubModules::Auth,
+                "executor" => SubModules::Executor,
+                _ => SubModules::Unknown,
+            }
+        } else {
+            SUBMODULES_UNKNOWN
+        }
     }
```

Comment:
A simple way to do this ：
https://play.rust-lang.org/?gist=5aca84a0e1c3e7bde3457d6bc696f0c8&version=stable

``` rust
match s.split('.').next().unwrap() {
    "a" => println!("got it"),
    "" => println!("empty"),
    _ => println!("others")
}
```